pkg/auth/authproviders/provider_commands.go

```diff
@@ -16,59 +16,88 @@ import (
 
 // DefaultAddToStore adds the providers stored data to the input store.
 func DefaultAddToStore(ctx context.Context, store Store) ProviderOption {
-	return func(pr *providerImpl) error {
+	return func(pr *providerImpl) (RevertOption, error) {
 		if pr.doNotStore {
-			return nil
+			return noOpRevert, nil
+		}
+		providerID := pr.storedInfo.GetId()
+		revert := func(pr *providerImpl) error {
+			return store.RemoveAuthProvider(ctx, providerID, true)
 		}
 		if pr.storedInfo.LastUpdated == nil {
 			pr.storedInfo.LastUpdated = protocompat.TimestampNow()
 		}
-		return store.AddAuthProvider(ctx, pr.storedInfo)
+		return revert, store.AddAuthProvider(ctx, pr.storedInfo)
 	}
 }
 
 // UpdateStore updates the stored value for the provider in the input store.
 func UpdateStore(ctx context.Context, store Store) ProviderOption {
-	return func(pr *providerImpl) error {
+	return func(pr *providerImpl) (RevertOption, error) {
 		if pr.doNotStore {
-			return nil
+			return noOpRevert, nil
+		}
+		providerID := pr.storedInfo.GetId()
+		oldProvider, found, err := store.GetAuthProvider(ctx, providerID)
```

Comment:
Let's decouple revert logic from forward logic (i.e. move into a dedicated function). It can be applied to all option functions.